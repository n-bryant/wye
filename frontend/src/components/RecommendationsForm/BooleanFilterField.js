import React from "react";
import PropTypes from "prop-types";

import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";
import { withStyles } from "@material-ui/core/styles";

import Switch from "@material-ui/core/Switch";
import Typography from "@material-ui/core/Typography";

import styles from "./BooleanFilterField.styles";

/**
 * renders a field for electing a boolean value in the form of a Switch
 */
export const BooleanFilterField = props => {
  const classnames = BooleanFilterField.classnames(props);
  const { classes, name, type, label, setFieldValue, value } = props;
  const [checked, setChecked] = React.useState(false);

  /**
   * handle changes to the Switch,
   * and update formik field value
   * @param {String} option
   */
  const handleChange = event => {
    setChecked(event.target.checked);
    setFieldValue(name, {
      ...value,
      gameFilters: {
        ...value.gameFilters,
        [type]: event.target.checked
      }
    });
  };

  return (
    <div className={classnames.root()}>
      <Typography
        className={classnames.element("label")}
        variant="body2"
        gutterBottom={true}
      >
        {label}
      </Typography>
      <div>
        <Switch
          classes={{
            switchBase: classes.switch,
            checked: classes.switchChecked,
            track: classes.switch
          }}
          checked={checked}
          onChange={handleChange}
          value={label}
          color="primary"
          inputProps={{ "aria-label": `${type}-checkbox` }}
        />
      </div>
    </div>
  );
};
BooleanFilterField.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}BooleanFilterField`
);
BooleanFilterField.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    label: PropTypes.string,
    switch: PropTypes.string,
    switchChecked: PropTypes.string
  }),
  // Formik field name
  name: PropTypes.string,
  // the label to display for the field
  label: PropTypes.string,
  // the type of field
  type: PropTypes.oneOf(["freeToPlay", "onSale"]).isRequired,
  // gameFilters field value from Formik form
  value: PropTypes.object,
  // handler for updating users field value
  setFieldValue: PropTypes.func
};
BooleanFilterField.defaultProps = {
  classes: {}
};

// apply styles
export default withStyles(styles)(BooleanFilterField);
