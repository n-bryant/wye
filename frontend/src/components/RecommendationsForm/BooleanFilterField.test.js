import Switch from "@material-ui/core/Switch";
import { BooleanFilterField } from "./BooleanFilterField";

describe("BooleanFilterField", () => {
  const props = {
    classes: {
      root: "root",
      label: "label",
      switch: "switch",
      switchChecked: "switchChecked"
    },
    name: "name",
    label: "label",
    type: "freeToPlay",
    value: {
      gameFilters: {}
    },
    setFieldValue: jest.fn()
  };

  let wrapper;
  beforeEach(() => {
    const checked = false;
    const setChecked = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(() => [checked, setChecked]);
    wrapper = shallow(<BooleanFilterField {...props} />);
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should update the formik field value when a change occurs", () => {
    wrapper.find(Switch).prop("onChange")({ target: { checked: true } });
    expect(props.setFieldValue).toHaveBeenCalledWith(props.name, {
      gameFilters: {
        [props.type]: true
      }
    });
  });
});
