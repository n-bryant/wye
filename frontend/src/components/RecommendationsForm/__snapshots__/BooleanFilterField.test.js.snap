// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooleanFilterField should render successfully 1`] = `
<div
  className="wye-BooleanFilterField root"
>
  <WithStyles(ForwardRef(Typography))
    className="wye-BooleanFilterField__label label"
    gutterBottom={true}
    variant="body2"
  >
    label
  </WithStyles(ForwardRef(Typography))>
  <div>
    <WithStyles(ForwardRef(Switch))
      checked={false}
      classes={
        Object {
          "checked": "switchChecked",
          "switchBase": "switch",
          "track": "switch",
        }
      }
      color="primary"
      inputProps={
        Object {
          "aria-label": "freeToPlay-checkbox",
        }
      }
      onChange={[Function]}
      value="label"
    />
  </div>
</div>
`;
